[C++ API] Add named submodule support to nn::Sequential

yf225 opened the pull request:
Previously, we were not able to assign names to `nn::Sequential`'s submodules. This PR adds this feature to match the Python API. Example use:
```cpp
Sequential sequential(named_submodule({
      {"linear", Linear(10, 3)},
      {"conv2d", Conv2d(1, 2, 3)},
      {"dropout", Dropout(0.5)},
      {"batchnorm", BatchNorm(5)},
      {"embedding", Embedding(4, 10)},
      {"lstm", LSTM(4, 5)}
}));
```

It also enables loading parameters of Python `nn.Sequential` module with custom submodules names into C++ frontend, unblocking https://github.com/pytorch/vision/pull/728#issuecomment-466661344.

ebetica commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -173,40 +173,50 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+  template <typename M>
+  void push_back(std::string name, M module) {
+    auto index = add_to_modules(module);
+    register_module(name, modules_[index].ptr());
```
nit: can `std::move(name)` here.

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -173,40 +173,50 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+
+  /// Adds a new named module to the `Sequential` container, with name of `std::string` type.
+  template <typename M>
+  void push_back(std::string name, M module) {
```
I don't understand why the public string parameter versions of push_back don't match the public non-string versions.  I would expect you would just have string overloads for the 3 existing ones?

ebetica commented on test/cpp/api/sequential.cpp:
```diff
@@ -64,6 +65,81 @@ TEST_F(SequentialTest, ConstructsFromModuleHolder) {
 }
 
+TEST_F(SequentialTest, ConstructsFromSharedPointerWithName) {
+  struct M : torch::nn::Module {
```
Why does this test duplicate so much code, won't a single definition of this M struct work?

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -173,40 +173,50 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+
+  /// Adds a new named module to the `Sequential` container, with name of `const char*` type.
+  template <typename M>
+  void push_back(const char* name, M module) {
```
why do you have both `string` and `const char*` versions of `push_back`?

ebetica replied:
I believe this does an extra copy of the module unnecessarily, and doesn't really follow the convention set above. I'm not sure how expensive the copy constructor is for Module.

It may be better to follow the construct given above, having a separate function for each of the possible types, i.e.
- std::shared_ptr<ModuleType>
- const ModuleHolder<M>&
- M&& with M being a module

Alternatively if you think the construct doesn't matter and the extra copy is whatever, then we can simplify the above code, but I think it's important to be consistent.

goldsborough replied:
I don't think we need `const char*` in the interface, since we eventually store `std::string`. It's ok to just have `std::string` as the argument type and move that properly into the eventual storage

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -173,40 +173,50 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+  void push_back(std::string name, M module) {
+    auto index = add_to_modules(module);
+    register_module(name, modules_[index].ptr());
+  }
```
why isn't there a constructor that takes an OrderedDict?  If I want to use names, I have to call `push_back`.  Also, this doesn't appear to document the behavior of combining named and unnamed modules.  What's the plan here?

gchanan commented on test/cpp/api/sequential.cpp:
```diff
@@ -64,6 +65,81 @@ TEST_F(SequentialTest, ConstructsFromModuleHolder) {
+
+  Sequential sequential(
+    std::make_shared<M>(1),
+    "m2", std::make_shared<M>(2),
```
this behavior is kind of crazy and doesn't match python.

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -311,15 +337,62 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+  template <typename M, typename = torch::detail::enable_if_module_t<M>>
+  size_t add_to_modules(M&& module) {
+    // Need to get rid of any reference components for make_unique.
+    using Type = typename std::remove_reference<M>::type;
```
I also don't quite get the point of `add_to_modules` (are you always properly moving/forwarding parameters around)?  Why don't the "final" non-string versions of `push_back` just call the string versions?

yf225 commented:
Note to self: make sure we are not making unnecessary copies (by adding tests for it)

goldsborough commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -298,6 +308,22 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+
+  /// Matches `Sequential("m1", Module(1), ...)` case
+  template <typename Module, typename... Rest>
+  void push_back(const char* name, Module&& module, Rest&&... rest) {
```
You can remove this overload and the whole `const char*` business, it's not worth it

goldsborough reviewed (changes requested):
I think @gchanan captures my opinion here in that the changes to `Sequential` are more complicated than they need to be. Essentially you should just need one overload for every existing `push_back` that has a string as the first parameter. The existing methods can just call into the new methods, using `modules_.size()` (the next index) as the name. I would do it like this:

1. Starting with the previous code, turn every push_back into a named version. All the logic should be in the named functions.
2. Thread the name through to the final insertion into the map
3. For every named version, add an unnamed overload (the previous signatures), but all they do is call `push_back(std::to_string(modules_.size()), <module>)`.

Then you effectively just have to add three methods and that's it

gchanan commented:
@goldsborough a few opinion questions for you:

1) what do you think about mixing named and "unnamed" parameters.  Should we allow it or not?

2) Having a constructor that takes an OrderedDict of modules would match the python interface.  But it seems like constructing an OrderedDict of modules by hand is annoying because of the different module types.  Should we have a helper that does it?  I guess you'd basically have to reproduce the module bits of Sequential here to do it though, i.e. three overloads: `shared_ptr<ModuleType>`, `ModuleHolder<M>`, `AnyModule`.

3) Is there documentation that  explains those three overloads anywhere?  I think I found some documentation around why there isn't a single module type (which makes sense), but I couldn't find docs about why those 3 are the "magic 3".

yf225 commented:
Another issue I found is that if I remove the copy constructor for the concrete module type test: https://github.com/pytorch/pytorch/blob/master/test/cpp/api/sequential.cpp#L38-L44, the test will fail to compile, because 
```
template <typename M, typename = torch::detail::enable_if_module_t<M>>
void push_back(M&& module) {
    // Need to get rid of any reference components for make_unique.
    using Type = typename std::remove_reference<M>::type;
    // Here we move (or copy) the module into a new shared_ptr.
    push_back(std::make_shared<Type>(std::forward<M>(module))); // NOTE: This line copies the module
}
```
(https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/modules/sequential.h#L202) actually copies the module and expects the copy constructor to exist. Ideally this copy should be avoided, and I am investigating how we can achieve so.

Commit:
allow copy only once

Commit:
simple ordereddict test

Commit:
try to simplify ordereddict creation

yf225 commented:
I updated the PR with a completely new design. Here is how I got here:

The original flat design is bad because it doesn't establish clear boundaries between different submodules. For example, `Sequential sequential("m1", M(1), "m2", M(2))` seems to suggest that `"m1"` and `M(1)` are objects of the same type, which is not true and can make people confused.

For constructor that takes an OrderedDict, (e.g. `Sequential sequential_named(OrderedDict({{"m1", M(1)}, {"m2", M(2)}}))`, technically speaking it's not possible to achieve this. For demonstration, I made an experimental branch at https://github.com/pytorch/pytorch/compare/master...yf225:cpp_sequential_named_submodules_ordereddict, which throws compile error saying that the type of template parameter `M` is not able to be deduced, and the reason is that brace-enclosed initializer list doesn't have a type inherently, and it's impossible for template argument deduction to figure out the type of `{"m2", M(2)}`'s inner members.

With this updated PR, I want to propose this API:
```
Sequential sequential_named(
    named_module("m1", M(1)),
    named_module(std::string("m2"), M(2)),
  );
```
`named_module()` is a helper function that constructs a `std::tuple` by using `std::forward_as_tuple()`, which preserves the value categories (aka. lvalue/rvalue) of the parameters passed in. In our case, it will make sure that we are constructing the tuple with `M(1)` / `M(2)` as rvalue references, and we are not making unnecessary copies of them when not needed (I added a test to ensure this).

One advantage of this API is that it looks similar to the unnamed module API, and users just need to substitute `M(1)` with `named_module("m1", M(1))` when they want to add a name to the submodule. Considering that OrderedDict is not possible to achieve, I do think that this is the best API we can add.

If we agree on this API design, then we can proceed to look at whether we are making the proper moving and forwarding, and we can discuss whether we should allow mixing named and unnamed submodules (which I think we shouldn't).

Commit:
[WIP]

Commit:
try helper method approach

Commit:
[WIP]

Commit:
better comments

Commit:
try something different

Commit:
Revert "try something different"

This reverts commit 848c7fcec70b30b0fa66b79ca700c2e9e84791ef.

Commit:
better comment

yf225 commented:
I figured out how to have a simple API for making OrderedDict work, and updated the PR to reflect the new approach.

With the new `named_submodules()` API, we will be creating an OrderedDict of named submodules much like how we do it with the Python API, which is much better for API parity.

yf225 commented on test/cpp/api/sequential.cpp:
```diff
@@ -32,20 +32,43 @@ TEST_F(SequentialTest, ConstructsFromSharedPointer) {
+    M(const M& other) : torch::nn::Module(other) {
+      // NOTE: The current implementation expects the module to be copied once
+      // when it's passed into `std::make_shared<T>()`.
+      // TODO: Find a way to avoid copying, and then delete the copy constructor.
```
I will file an issue to document this problem.

Commit:
better comments

yf225 commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -90,6 +90,66 @@ namespace nn {
 /// \endrst
 class SequentialImpl : public Cloneable<SequentialImpl> {
  public:
+  class NamedSubmodule {
```
For why we need to have this class, see https://github.com/pytorch/pytorch/pull/17552/files#diff-ad62a46b3133f7412c5667d4c194359eR395.

Commit:
don't std::move named_module.name()

Commit:
fix lint

gchanan commented on test/cpp/api/sequential.cpp:
```diff
@@ -32,20 +32,43 @@ TEST_F(SequentialTest, ConstructsFromSharedPointer) {
+      // NOTE: The current implementation expects the module to be copied once
+      // when it's passed into `std::make_shared<T>()`.
+      // TODO: Find a way to avoid copying, and then delete the copy constructor.
+      AT_ASSERT(copy_count == 0);
```
this always has to be true, right?

yf225 replied:
yes this is only true when the copy constructor is called the first time, since we increment `copy_count` after this line. When the problem mentioned at https://github.com/pytorch/pytorch/issues/17879 is fixed, we will be able to remove the copy constructor completely and don't need to do this check anymore.

gchanan replied:
how can the copy constructor of an object be called multiple times?

yf225 replied:
It was an oversight, I fixed it now with a global variable that stores the copy count.

gchanan commented on test/cpp/api/sequential.cpp:
```diff
@@ -81,6 +113,32 @@ TEST_F(SequentialTest, PushBackAddsAnElement) {
+  ASSERT_EQ(sequential_named->size(), 2);
+  ASSERT_EQ(sequential_named->named_children()[1].key(), "linear2");
+
+  sequential_named->push_back({"shared_m1", std::make_shared<M>(1)});
```
are there any tests for mixing named and unnamed push_backs?  What do you expect to happen there?

yf225 replied:
I updated it to test mixing named and unnamed push_backs, it should just work as expected.

gchanan replied:
I don't see where you actually check the names of the unnamed push_backs.

yf225 replied:
Fixed

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -90,6 +90,66 @@ namespace nn {
 /// \endrst
 class SequentialImpl : public Cloneable<SequentialImpl> {
  public:
+  class NamedSubmodule {
```
why do we need NamedSubmodule to be an internal type of Sequential as opposed to just some helper that creates OrderedDicts?

yf225 replied:
`NamedSubmodule` doesn't have to be an internal type of Sequential, but it has to be a class type so that we can use its constructors to resolve the types of elements in braced-init lists such as `{"m1", M(1)}`. If `NamedSubmodule()` is just a pure helper function, it won't be able to understand the type of `{"m1", M(1)}`, since braced-init list doesn't inherently have a type.

yf225 replied:
The comments here might be helpful: https://github.com/pytorch/pytorch/pull/17552/files#diff-ad62a46b3133f7412c5667d4c194359eR395

Commit:
try to fix Doxygen

Commit:
add test for mixing named and unnamed submodules

Commit:
add name check for unnamed submodules

Commit:
Merge branch 'master' of https://github.com/yf225/pytorch into cpp_sequential_named_submodules_split

Commit:
try to fix copy_count

Commit:
more fixes

Commit:
more fixes

Commit:
better comments

Commit:
clean up

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -90,6 +90,66 @@ namespace nn {
+      static_assert(
+          torch::detail::has_forward<ModuleType>::value,
+          "Can only add modules with a forward() method to Sequential");
+      new (this) NamedSubmodule(std::move(name), AnyModule(std::move(module_ptr)));
```
you don't want to do this -- it's extremely fragile.  Think about what happens if you allocate memory earlier or if we ever subclass these.

just use NamedSubmodule(...):NamedSubmodule(), which should be fine, because all the code here is static_asserts.

You have this a few places.

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -196,17 +245,30 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+  /// Adds a new `NamedSubmodule` to the `Sequential` container.
+  /// This method allows creating the `NamedSubmodule` from braced-init-list, e.g.
+  /// `push_back({"linear", Linear(3, 4)})`
+  void push_back(NamedSubmodule named_module) {
```
can you not provide push_backs for (string, module)?  NamedSubmodule being part of the public interface feels clunky, because what you get back from named_submodule isn't a NamedSubmodule at all!  It's just an OrderedDict of modules, which you can handle perfectly well.

Commit:
refactor

yf225 commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -416,10 +416,19 @@ struct AnyModule::Holder : public AnyModule::Placeholder {
+      "Can only store object derived from nn::Module into AnyModule");
+  static_assert(
+      torch::detail::has_forward<ModuleType>::value,
+      "Can only store module with a forward() method into AnyModule");
```
We move the `static_assert`s here because `AnyModule` needs to uphold the expectation of storing only `nn::Module` subclass object that provides a `forward()` method.

gchanan replied:
can't the stuff below still be done as member-wise initialization?  Because the only stuff here are static_asserts.

yf225 commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -175,18 +175,7 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
-    // templatized, so it'll give a nasty compiler error.
-    static_assert(
-        !std::is_same<SequentialImpl, ModuleType>::value,
-        "Sequential is not nestable");
```
With the new design we are not able to do this `static_assert` at compile time, because `NamedAnyModule` has no knowledge of Sequential when it unwraps the `std::shared_ptr<ModuleType>` at https://github.com/pytorch/pytorch/pull/17552/files#diff-3ce0ff9395bf23f007205836f43d31aaR573. However, we can safely remove this `static_assert` because it will be caught by https://github.com/pytorch/pytorch/pull/17552/files#diff-3ce0ff9395bf23f007205836f43d31aaR425.

gchanan replied:
I don't see how this would be caught at the linked line -- there is no code about Sequential there.

yf225 replied:
With the change in this PR, if we run the following test case:
```cpp
struct M : torch::nn::Module {
  explicit M(int value_) : value(value_) {}
  int value;
  int forward() {
    return value;
  }
};

Sequential inner_sequential(M(1));
Sequential outer_sequential(inner_sequential, inner_sequential);  // nested nn::Sequential

ASSERT_EQ(outer_sequential->size(), 2);
```
The compiler will give an error:
```
../torch/csrc/api/include/torch/nn/modules/any.h:422:61: error: no matching function for call to ‘torch::nn::AnyModule::make_holder(std::remove_reference<std::shared_ptr<torch::nn::SequentialImpl>&>::type, <unresolved overloaded function type>)’
           &std::remove_reference<ModuleType>::type::forward)) {
                                                             ^
../torch/csrc/api/include/torch/nn/modules/any.h:526:41: note: candidate: template<class ModuleType, class Class, class ReturnType, class ... ArgumentTypes> std::unique_ptr<torch::nn::AnyModule::Placeholder> torch::nn::AnyModule::make_holder(std::shared_ptr<_Tp1>&&, ReturnType (Class::*)(ArgumentTypes ...))
 std::unique_ptr<AnyModule::Placeholder> AnyModule::make_holder(
                                         ^
../torch/csrc/api/include/torch/nn/modules/any.h:526:41: note:   template argument deduction/substitution failed:
../torch/csrc/api/include/torch/nn/modules/any.h:422:61: note:   couldn't deduce template parameter ‘Class’
           &std::remove_reference<ModuleType>::type::forward)) {
                                                             ^
../torch/csrc/api/include/torch/nn/modules/any.h:428:3: error: static assertion failed: Can only store module with a forward() method into AnyModule
   static_assert(
   ^
```
The compiler throws this error because it isn't able to deduce the return type of the inner `nn::Sequential`'s `forward()` method, since it's templated: https://github.com/pytorch/pytorch/blob/50df3e5e2e2c51db6389b5c47c048f9db112c5c3/torch/csrc/api/include/torch/nn/modules/sequential.h#L152-L153. This compiler complaint is exactly what we want, because we don't want to allow nesting `nn::Sequential`.

The only downside to this approach is that the error message is not as informative (i.e. it only says "Can only store module with a forward() method into AnyModule", but the real reason is that "Sequential is not nestable"). An alternative way to have a better error message is to keep the `!std::is_same<SequentialImpl, ModuleType>::value` check here, and add a runtime check at https://github.com/pytorch/pytorch/pull/17552/files#diff-ad62a46b3133f7412c5667d4c194359eR340, to prevent the user from pushing in `nn::Sequential` through either the direct push_back method, or the ordered_dict method (which doesn't go through this path).

yf225 commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -175,18 +175,7 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
-    static_assert(
-        torch::detail::has_forward<ModuleType>::value,
-        "Can only add modules with a forward() method to Sequential");
-    push_back(AnyModule(std::move(module_ptr)));
```
These two are moved to https://github.com/pytorch/pytorch/pull/17552/files#diff-3ce0ff9395bf23f007205836f43d31aaR422.

yf225 commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -175,18 +175,7 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
-        "Can only add objects derived from nn::Module to Sequential");
-    static_assert(
-        torch::detail::has_forward<ModuleType>::value,
-        "Can only add modules with a forward() method to Sequential");
```
These two are moved to https://github.com/pytorch/pytorch/pull/17552/files#diff-3ce0ff9395bf23f007205836f43d31aaR422.

yf225 commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -196,17 +185,23 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
-    // Need to get rid of any reference components for make_unique.
-    using Type = typename std::remove_reference<M>::type;
-    // Here we move (or copy) the module into a new shared_ptr.
-    push_back(std::make_shared<Type>(std::forward<M>(module)));
```
The logic here is moved to https://github.com/pytorch/pytorch/pull/17552/files#diff-3ce0ff9395bf23f007205836f43d31aaR580.

Commit:
Merge branch 'master' of https://github.com/yf225/pytorch into cpp_sequential_named_submodules_split

Commit:
Merge branch 'master' of https://github.com/yf225/pytorch into cpp_sequential_named_submodules_split

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -175,18 +175,7 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
-        torch::detail::has_forward<ModuleType>::value,
-        "Can only add modules with a forward() method to Sequential");
-    push_back(AnyModule(std::move(module_ptr)));
+    push_back(modules_ordered_dict({{std::to_string(modules_.size()), std::move(module_ptr)}}));
```
I don't understand why sequential needs to know anything about modules_ordered_dict or named modules -- can you not just add new overloads for all existing `push_back`s, move all the checking to those functions, and just have the existing functions forward to the new ones, passing the right name?  And an OrderedDict constructor overload that extracts the modules and checks them in the standard way.

yf225 replied:
Fixed

Commit:
clean up modules_ordered_dict

Commit:
use new push_back API

yf225 commented on test/cpp/api/sequential.cpp:
```diff
@@ -32,20 +32,47 @@ TEST_F(SequentialTest, ConstructsFromSharedPointer) {
 }
 
 TEST_F(SequentialTest, ConstructsFromConcreteType) {
+  static int copy_count;
```
Ideally we want the `copy_count` to be inside the struct as a static member (so that it can keep track of the number of copy constructor calls happened for this class), however static data members are not allowed in local classes (https://stackoverflow.com/questions/8162994/why-arent-static-data-members-allowed-in-local-classes), and we have to declare it as a global static variable instead.

gchanan replied:
is it global static or static local variable?  Also, that answered why it isn't static in the struct but why is it static at all?

yf225 replied:
`copy_count` is within the scope of `SequentialTest::ConstructsFromConcreteType`, and won't be polluting other tests. If we define it as a non-static variable, it will throw `error: use of local variable with automatic storage from containing function` error (https://stackoverflow.com/questions/43932288/declared-class-inside-the-main-function-in-c), and the workaround is to either define it as a static variable (which is the chosen approach here) or pass it into the copy constructor.

gchanan replied:
ah I see, makes sense.

Commit:
Revert "use new push_back API"

This reverts commit 4c0eab960c3a06ead56f2cb71fa3e5b21e18f25a.

Commit:
make copy_count non-static

Commit:
Revert "make copy_count non-static"

This reverts commit ae816c751a56f1d7ad30401d5b3bcdab3670b148.

Commit:
use new push_back API

Commit:
clean up and comment

gchanan commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -196,17 +191,39 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+  /// Adds an `OrderedDict` of named `AnyModule`s to the `Sequential` container.
+  /// Combining with `modules_ordered_dict()`, it enables the following use case:
+  /// `Sequential sequential(modules_ordered_dict({{"m1", M(1)}, {"m2", M(2)}}))`
+  void push_back(torch::OrderedDict<std::string, AnyModule>&& ordered_dict) {
```
why is this a `push_back` and not a constructor?

yf225 replied:
Fixed

gchanan commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +561,65 @@ ModuleType& AnyModule::get_(
+
+  /// Creates a `NamedAnyModule` from a (boxed) `Module`.
+  template <typename ModuleType>
+  NamedAnyModule(std::string name, std::shared_ptr<ModuleType> module_ptr) : NamedAnyModule(std::move(name), AnyModule(std::move(module_ptr))) {}
```
can you format this nicer?  Look at the entire file -- it's quite obvious what part you wrote :).

yf225 replied:
Fixed

gchanan commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +561,65 @@ ModuleType& AnyModule::get_(
+inline torch::OrderedDict<std::string, AnyModule> modules_ordered_dict(std::initializer_list<NamedAnyModule> named_modules) {
+  torch::OrderedDict<std::string, AnyModule> dict;
+  for (auto named_module : named_modules) {
+    dict.insert(named_module.name(), std::move(named_module.module()));
```
I guess I have a more basic question -- why do I need an initializer list here at all?  For example, SequentialImpl can be called via {...} but doesn't have to deal with initializer lists.

yf225 replied:
Is there a concrete example for calling SequentialImpl via {...}? I think the reason here is that we can generally call `Class object{a, b}` if there is a constructor in `Class` that matches the signature `(TypeA, TypeB)`. However, nested braced-init list such as `Class object{{"m1", M(1)}, {"m2", M(2)}}` will not work because there will never be a type that can match to `{"m1", M(1)}` or `{"m2", M(2)}`, since they actually don't have a type at all.

Commit:
move content_ to member-wise initialization

Commit:
Merge branch 'master' of https://github.com/yf225/pytorch into cpp_sequential_named_submodules_split

Commit:
better formatting for any.h

Commit:
better comment

Commit:
nit

Commit:
move to constructor

Commit:
fix formatting

goldsborough commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
+class NamedAnyModule {
+ public:
+  /// We don't allow constructing an empty `NamedAnyModule`.
+  NamedAnyModule() = delete;
```
You don't really need this one. If there are non-default constructors, the default one is always disabled. You can keep it for clarity if you want

goldsborough commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
+      : NamedAnyModule(std::move(name), module_holder.ptr()) {}
+
+  /// Returns a reference to the name.
+  std::string name() const noexcept {
```
You are returning a copy each time here. You most likely want `const std::string&` return type. As an aside: If you look at https://en.cppreference.com/w/cpp/string/basic_string/basic_string, you'll see that the copy constructor of `std::string` is not `noexcept`. So the `noexcept` here would be wrong. It will be correct if you change the return type to `const std::string&`

goldsborough commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ NamedAnyModule ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+class NamedAnyModule {
```
Please add some high level docs to this class to understand why it exists

goldsborough replied:
I think you should create a separate file for this. 600 lines is pretty long for a file and not everyone who needs `AnyModule` will need `NamedAnyModule`.

goldsborough commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
+
+ private:
+  /// Creates a `NamedAnyModule` from a type-erased `AnyModule`.
+  NamedAnyModule(std::string name, AnyModule any_module) {
```
Please use the initializer list. `NamedAnyModule(std::string name, AnyModule any_module) : name_(std::move(name)), module_(std::move(any_module) {}`

goldsborough commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
+// `modules_ordered_dict({{"m1", M(1)}, {"m2", M(2)}})`,
+// if we use the second signature, at the template argument deduction step
+// the compiler is not able to deduce the type of `ModuleType` to the type of
+// `M(1)` or `M(2)`, since the compiler doesn't actually look into the
```
Did you try whether `std::pair<std::string, AnyModule>` works?

yf225 replied:
I tried to change the `modules_ordered_dict(...)` function to ```torch::OrderedDict<std::string, AnyModule> modules_ordered_dict(std::initializer_list<std::pair<std::string, AnyModule>> named_modules)```, however it doesn't seem to work and throws:
```cpp
../test/cpp/api/sequential.cpp: In member function ‘virtual void SequentialTest_ConstructsFromConcreteType_Test::TestBody()’:
../test/cpp/api/sequential.cpp:72:4: error: could not convert ‘{{"m1", SequentialTest_ConstructsFromConcreteType_Test::TestBody()::M(1)}, {std::__cxx11::basic_string<char>(((const char*)"m2"), std::allocator<char>()), SequentialTest_ConstructsFromConcreteType_Test::TestBody()::M(2)}, {"m3", SequentialTest_ConstructsFromConcreteType_Test::TestBody()::M(3)}}’ from ‘<brace-enclosed initializer list>’ to ‘std::initializer_list<std::pair<std::__cxx11::basic_string<char>, torch::nn::AnyModule> >’
```
The compiler is not able to match `std::initializer_list<std::pair<std::string, AnyModule>>` to the nested braced-init list `{{"m1", M(1)}, {std::string("m2"), M(2)}, {"m3", M(3)}}`. So I think the NamedAnyModule approach here is necessary.

goldsborough commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
+// the `NamedAnyModule` constructors, and let the constructors do the job of
+// figuring out the types of its elements and do the matching to the correct
+// module type.
+inline torch::OrderedDict<std::string, AnyModule> modules_ordered_dict(
```
My personal preference was always that if it's not a template method, the definition should go into a .cpp file.

goldsborough commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
+inline torch::OrderedDict<std::string, AnyModule> modules_ordered_dict(
+  std::initializer_list<NamedAnyModule> named_modules) {
+  torch::OrderedDict<std::string, AnyModule> dict;
+  for (auto named_module : named_modules) {
```
Note that you're making copies here?

yf225 replied:
I think we need to make `named_module` non-const one way or another, because `std::initializer_list<T>` only provides access to an array of objects of type `const T`, not `T` (according to https://en.cppreference.com/w/cpp/utility/initializer_list), but we need `named_module` to be of non-const type to be able to do `std::move(named_module.module())`. 

In the latest commit, I changed `std::move(named_module.module())` to `std::move(const_cast<NamedAnyModule&>(named_module).module()))` so that we can avoid doing copies in this line.

goldsborough commented on torch/csrc/api/include/torch/nn/modules/any.h:
```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
+// figuring out the types of its elements and do the matching to the correct
+// module type.
+inline torch::OrderedDict<std::string, AnyModule> modules_ordered_dict(
+  std::initializer_list<NamedAnyModule> named_modules) {
```
You're using `std::initializer_list`, but I don't see an include for `<initializer_list>`

goldsborough commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -302,7 +320,12 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
+  /// or `push_back("name", module)`, since they should be handled by their respective
+  /// `push_back` functions.
+  template <typename First, typename Second, typename... Rest,
+    typename = torch::disable_if_t<std::is_same<First, std::string>::value ||
```
Why do we need the `const char*` guard? I think we're only using `std::string` in here?

yf225 replied:
If we don't add the guard for `const char*` here, a call such as `sequential->push_back("shared_m1", M(1))` will be template-matched to this `push_back(First&& first, Second&& second, Rest&&... rest)` method, which is not what we want (we want it to match to `push_back(std::string name, M&& module)` instead.

goldsborough commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -312,9 +335,13 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
 
   /// Adds a type-erased `AnyModule` to the `Sequential`.
   void push_back(AnyModule any_module) {
```
Is this method still used now?

yf225 replied:
It is still used by this line: https://github.com/pytorch/pytorch/blob/c3e3c5cc39165470ddab5afb6373399fdbd6598e/torch/csrc/api/include/torch/nn/modules/sequential.h#L111

goldsborough commented on torch/csrc/api/include/torch/nn/modules/sequential.h:
```diff
@@ -312,9 +335,13 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
     modules_.push_back(std::move(any_module));
     const auto index = modules_.size() - 1;
-    register_module(std::to_string(index), modules_[index].ptr());
+    register_module(name, modules_[index].ptr());
```
`std::move(name)`?

goldsborough reviewed (commented):
I left some comments

Commit:
fixes

Commit:
Merge branch 'master' of https://github.com/yf225/pytorch into cpp_sequential_named_submodules_split

Commit:
fix

Commit:
const_cast

Commit:
named_any

Commit:
add docs for NamedAnyModule

Commit:
try to fix Windows build error

goldsborough reviewed (approved):
👍

Commit:
const_cast is not allowed by clang-tidy

Commit:
fix cpp doc check
